File: multiCamCommon/swift/Sources/MultiCamCommon/FileTypes.swift
import Foundation

/// Metadata for a single video file
public struct FileMetadata: Codable {
    /// Filename
    public let fileName: String

    /// File size in bytes
    public let fileSize: Int64

    /// File creation time (Unix timestamp)
    public let creationDate: TimeInterval

    /// File modification time (Unix timestamp)
    public let modificationDate: TimeInterval

    public init(
        fileName: String,
        fileSize: Int64,
        creationDate: TimeInterval,
        modificationDate: TimeInterval
    ) {
        self.fileName = fileName
        self.fileSize = fileSize
        self.creationDate = creationDate
        self.modificationDate = modificationDate
    }
}

/// Header for binary file transfer
///
/// This is sent as JSON before the binary file data in GET_VIDEO responses.
///
/// Binary protocol:
/// 1. Header size (4 bytes, big-endian uint32)
/// 2. JSON FileResponse header
/// 3. Binary file data
public struct FileResponse: Codable {
    /// Device that owns the file
    public let deviceId: String

    /// Filename
    public let fileName: String

    /// File size in bytes
    public let fileSize: Int64

    /// Status (typically "ready")
    public let status: String

    public init(
        deviceId: String,
        fileName: String,
        fileSize: Int64,
        status: String
    ) {
        self.deviceId = deviceId
        self.fileName = fileName
        self.fileSize = fileSize
        self.status = status
    }

    // MARK: - Serialization

    /// Serialize file response to JSON data
    /// - Returns: JSON data
    /// - Throws: Encoding error
    public func toJSON() throws -> Data {
        let encoder = JSONEncoder()
        return try encoder.encode(self)
    }

    /// Deserialize file response from JSON data
    /// - Parameter data: JSON data
    /// - Returns: FileResponse instance
    /// - Throws: Decoding error
    public static func fromJSON(_ data: Data) throws -> FileResponse {
        let decoder = JSONDecoder()
        return try decoder.decode(FileResponse.self, from: data)
    }
}

/// Response to LIST_FILES command
///
/// Note: This command may not be supported on all platforms.
public struct ListFilesResponse: Codable {
    /// Device ID
    public let deviceId: String

    /// Status (see DeviceStatus enum)
    public let status: String

    /// Response timestamp
    public let timestamp: TimeInterval

    /// List of available files
    public let files: [FileMetadata]

    public init(
        deviceId: String,
        status: String,
        timestamp: TimeInterval,
        files: [FileMetadata]
    ) {
        self.deviceId = deviceId
        self.status = status
        self.timestamp = timestamp
        self.files = files
    }

    // MARK: - Serialization

    /// Serialize list files response to JSON data
    /// - Returns: JSON data
    /// - Throws: Encoding error
    public func toJSON() throws -> Data {
        let encoder = JSONEncoder()
        return try encoder.encode(self)
    }

    /// Deserialize list files response from JSON data
    /// - Parameter data: JSON data
    /// - Returns: ListFilesResponse instance
    /// - Throws: Decoding error
    public static func fromJSON(_ data: Data) throws -> ListFilesResponse {
        let decoder = JSONDecoder()
        return try decoder.decode(ListFilesResponse.self, from: data)
    }
}

/// Response to STOP_RECORDING command
public struct StopRecordingResponse: Codable {
    /// Device ID
    public let deviceId: String

    /// Device status (typically "recording_stopped")
    public let status: String

    /// Response timestamp
    public let timestamp: TimeInterval

    /// File name of the recorded video
    public let fileName: String

    /// File size in bytes
    public let fileSize: Int64

    public init(
        deviceId: String,
        status: String,
        timestamp: TimeInterval,
        fileName: String,
        fileSize: Int64
    ) {
        self.deviceId = deviceId
        self.status = status
        self.timestamp = timestamp
        self.fileName = fileName
        self.fileSize = fileSize
    }

    // MARK: - Serialization

    /// Serialize response to JSON data
    /// - Returns: JSON data
    /// - Throws: Encoding error
    public func toJSON() throws -> Data {
        let encoder = JSONEncoder()
        return try encoder.encode(self)
    }

    /// Deserialize response from JSON data
    /// - Parameter data: JSON data
    /// - Returns: StopRecordingResponse instance
    /// - Throws: Decoding error
    public static func fromJSON(_ data: Data) throws -> StopRecordingResponse {
        let decoder = JSONDecoder()
        return try decoder.decode(StopRecordingResponse.self, from: data)
    }
}

/// Error response from a MultiCam device
public struct ErrorResponse: Codable {
    /// Device ID
    public let deviceId: String

    /// Error status (e.g., "file_not_found", "error")
    public let status: String

    /// Response timestamp
    public let timestamp: TimeInterval

    /// Human-readable error message
    public let message: String

    public init(
        deviceId: String,
        status: String,
        timestamp: TimeInterval,
        message: String
    ) {
        self.deviceId = deviceId
        self.status = status
        self.timestamp = timestamp
        self.message = message
    }

    // MARK: - Serialization

    /// Serialize response to JSON data
    /// - Returns: JSON data
    /// - Throws: Encoding error
    public func toJSON() throws -> Data {
        let encoder = JSONEncoder()
        return try encoder.encode(self)
    }

    /// Deserialize response from JSON data
    /// - Parameter data: JSON data
    /// - Returns: ErrorResponse instance
    /// - Throws: Decoding error
    public static func fromJSON(_ data: Data) throws -> ErrorResponse {
        let decoder = JSONDecoder()
        return try decoder.decode(ErrorResponse.self, from: data)
    }
}

File: multiCamCommon/java/src/main/java/com/multicam/common/FileTypes.java
package com.multicam.common;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import java.util.List;
import java.util.ArrayList;

/**
 * File-related data types for MultiCam API.
 */
public class FileTypes {
    private static final Gson gson = new GsonBuilder().create();

    /**
     * Metadata for a single video file.
     */
    public static class FileMetadata {
        /** Filename */
        public String fileName;

        /** File size in bytes */
        public long fileSize;

        /** File creation time (Unix timestamp) */
        public double creationDate;

        /** File modification time (Unix timestamp) */
        public double modificationDate;

        public FileMetadata() {
        }

        public FileMetadata(String fileName, long fileSize,
                          double creationDate, double modificationDate) {
            this.fileName = fileName;
            this.fileSize = fileSize;
            this.creationDate = creationDate;
            this.modificationDate = modificationDate;
        }
    }

    /**
     * Header for binary file transfer.
     * <p>
     * This is sent as JSON before the binary file data in GET_VIDEO responses.
     * <p>
     * Binary protocol:
     * <ol>
     *   <li>Header size (4 bytes, big-endian uint32)</li>
     *   <li>JSON FileResponse header</li>
     *   <li>Binary file data</li>
     * </ol>
     */
    public static class FileResponse {
        /** Device that owns the file */
        public String deviceId;

        /** Filename */
        public String fileName;

        /** File size in bytes */
        public long fileSize;

        /** Status (typically "ready") */
        public String status;

        public FileResponse() {
        }

        public FileResponse(String deviceId, String fileName,
                          long fileSize, String status) {
            this.deviceId = deviceId;
            this.fileName = fileName;
            this.fileSize = fileSize;
            this.status = status;
        }

        /**
         * Serialize file response to JSON string.
         *
         * @return JSON string representation
         */
        public String toJson() {
            return gson.toJson(this);
        }

        /**
         * Deserialize file response from JSON string.
         *
         * @param json JSON string to parse
         * @return FileResponse instance
         */
        public static FileResponse fromJson(String json) {
            return gson.fromJson(json, FileResponse.class);
        }
    }

    /**
     * Response to LIST_FILES command.
     * <p>
     * Note: This command may not be supported on all platforms.
     */
    public static class ListFilesResponse {
        /** Device ID */
        public String deviceId;

        /** Status (see DeviceStatus enum) */
        public String status;

        /** Response timestamp */
        public double timestamp;

        /** List of available files */
        public List<FileMetadata> files;

        public ListFilesResponse() {
            this.files = new ArrayList<>();
        }

        public ListFilesResponse(String deviceId, String status, double timestamp,
                               List<FileMetadata> files) {
            this.deviceId = deviceId;
            this.status = status;
            this.timestamp = timestamp;
            this.files = files != null ? files : new ArrayList<>();
        }

        /**
         * Serialize list files response to JSON string.
         *
         * @return JSON string representation
         */
        public String toJson() {
            return gson.toJson(this);
        }

        /**
         * Deserialize list files response from JSON string.
         *
         * @param json JSON string to parse
         * @return ListFilesResponse instance
         */
        public static ListFilesResponse fromJson(String json) {
            return gson.fromJson(json, ListFilesResponse.class);
        }
    }

    /**
     * Response to STOP_RECORDING command.
     */
    public static class StopRecordingResponse {
        /** Device ID */
        public String deviceId;

        /** Device status (typically "recording_stopped") */
        public String status;

        /** Response timestamp */
        public double timestamp;

        /** File name of the recorded video */
        public String fileName;

        /** File size in bytes */
        public long fileSize;

        public StopRecordingResponse() {
        }

        public StopRecordingResponse(String deviceId, String status, double timestamp,
                                    String fileName, long fileSize) {
            this.deviceId = deviceId;
            this.status = status;
            this.timestamp = timestamp;
            this.fileName = fileName;
            this.fileSize = fileSize;
        }

        /**
         * Serialize response to JSON string.
         *
         * @return JSON string representation
         */
        public String toJson() {
            return gson.toJson(this);
        }

        /**
         * Deserialize response from JSON string.
         *
         * @param json JSON string to parse
         * @return StopRecordingResponse instance
         */
        public static StopRecordingResponse fromJson(String json) {
            return gson.fromJson(json, StopRecordingResponse.class);
        }
    }

    /**
     * Error response from a MultiCam device.
     */
    public static class ErrorResponse {
        /** Device ID */
        public String deviceId;

        /** Error status (e.g., "file_not_found", "error") */
        public String status;

        /** Response timestamp */
        public double timestamp;

        /** Human-readable error message */
        public String message;

        public ErrorResponse() {
        }

        public ErrorResponse(String deviceId, String status, double timestamp, String message) {
            this.deviceId = deviceId;
            this.status = status;
            this.timestamp = timestamp;
            this.message = message;
        }

        /**
         * Serialize response to JSON string.
         *
         * @return JSON string representation
         */
        public String toJson() {
            return gson.toJson(this);
        }

        /**
         * Deserialize response from JSON string.
         *
         * @param json JSON string to parse
         * @return ErrorResponse instance
         */
        public static ErrorResponse fromJson(String json) {
            return gson.fromJson(json, ErrorResponse.class);
        }
    }
}
